File: internal/ui/characters.go
package ui

import (
	. "github.com/vipmax/edgo/internal/operations"
	. "github.com/vipmax/edgo/internal/utils"
)

func (e *Editor) AddChar(ch rune) {
	if len(e.Selection.GetSelectionString(e.code.Value())) != 0 { e.Cut(false) }

	e.Focus()
	e.InsertCharacter(e.Row, e.Col, ch)
	e.Col++

	val, found := e.MaybeAddPair(e.Row, e.Col, ch)
	if found {
		e.InsertCharacter(e.Row, e.Col, val)
	}

	e.Update = true
	e.UpdateLsp(false, string(e.code.Value()))
	e.IsContentChanged = true
	e.FindTests()
}

func (e *Editor) InsertCharacter(line, pos int, ch rune) {
	offset := LineOffset(e.code.Value(), line) + pos

	// adding one character offset after first line
	// this is required due to the newline characters in the string buffer
	if line > 0 { offset += 1 }

	e.code.Insert(offset, []byte(string(ch)))
	e.treeSitterHighlighter.InsertTextEdit(e.code.Value(), offset, len(string(ch)))

	// undo operation type is switched between insert and addchar types
	var op Action

	// change op if this is a new line character
	if ch == '\n' { op = Insert } else { op = AddChar }

	// record the operation on the undo stack. Note that we're creating a new EditOperation
	// and adding all the Operations to it
	e.Undo = append(e.Undo, EditOperation{{op, []byte(string(ch)), offset, CursorMove{line, pos}}})
}

func (e *Editor) InsertString(line, pos int, linestring string) {
	// modify string to remove tabs and space from beginning
	l := RemoveLeadingTabsSpaces(linestring)
	offset := LineOffset(e.code.Value(), line) + pos + 1
	e.code.Insert(offset, []byte(l))
	e.treeSitterHighlighter.InsertTextEdit(e.code.Value(), offset, len(l))

	// record the operation on the undo stack. Note that we're creating a new EditOperation
	// and adding all the Operations to it
	e.Undo = append(e.Undo, EditOperation{{Insert, []byte(l), offset, CursorMove{line, pos}}})
}

func (e *Editor) DeleteCharacter(line, pos int) {
	offset := LineOffset(e.code.Value(), line) + pos
	ch := e.code.At(offset)
	e.code.Remove(offset, offset + 1)
	e.treeSitterHighlighter.RemoveTextEdit(e.code.Value(), offset, len(string(ch)))

	// record the operation on the undo stack. Note that we're creating a new EditOperation
	// and adding all the Operations to it
	e.Undo = append(e.Undo, EditOperation{{Delete, []byte{ch}, offset, CursorMove{line, pos}}})
}

func (e *Editor) ReplaceString(line, from, end int, instext string) {
	offset := LineOffset(e.code.Value(), line)
	begin_idx := offset + from + 1
	end_idx := offset + end + 1

	deltext := make([]byte, end_idx - begin_idx)
	copy(deltext, e.code.Slice(begin_idx, end_idx))

	e.code.Remove(begin_idx, end_idx)
	e.treeSitterHighlighter.RemoveTextEdit(e.code.Value(), begin_idx, len(deltext))

	e.code.Insert(begin_idx, []byte(instext))
	e.treeSitterHighlighter.InsertTextEdit(e.code.Value(), begin_idx, len(instext))

	// record the operation on the undo stack. Note that we're creating a new EditOperation
	// and adding all the Operations to it
	e.Undo = append(e.Undo, EditOperation{
		{Delete, []byte(deltext), begin_idx, CursorMove{line, from}},
		{Insert, []byte(instext), begin_idx, CursorMove{line, from}},
	})
}

func (e *Editor) ShiftWithTabsToRight(line, pos int, selectedLines []int) {
	e.Selection.Ssx = 0

	var ops = EditOperation{}
	for _, linenumber := range selectedLines {
		offset := LineOffset(e.code.Value(), linenumber)
		e.code.Insert(offset, []byte("\t"))
		ops = append(ops, Operation{Insert, []byte("\t"), offset, CursorMove{line, pos}})
	}

	e.Selection.Sex = pos
	e.Undo = append(e.Undo, ops)
}

func (e *Editor) MaybeAddPair(line, pos int, ch rune) (rune, bool) {
	pairMap := map[rune]rune{
		'(': ')', '{': '}', '[': ']',
		'"': '"', '\'': '\'', '`': '`',
	}

	if e.code.Len() == 0 { return rune('\000'), false }

	offset_start := LineOffset(e.code.Value(), line)
	offset_end := LineOffset(e.code.Value(), line + 1)
	line_str := e.code.Slice(offset_start + 1, offset_end)
	current_char := e.code.At(offset_start + 1 + pos - 2)

	if closeChar, found := pairMap[ch]; found {
		noMoreChars := pos >= len(line_str) - 1
		isSpaceNext := pos < len(line_str) - 1 && current_char == ' '
		isStringAndClosedBracketNext := closeChar == '"' && pos < len(line_str) - 1 && current_char == ')'

		if noMoreChars || isSpaceNext || isStringAndClosedBracketNext {
			return closeChar, found
		}
	}

	return rune('\000'), false
}

File: source/characters.cpp
/**
    Copyright (C) 2023 - 2026, edgo authors

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, see <https://www.gnu.org/licenses/>.
*/

#include "editor.h"
#include "operations.h"
#include "utils.h"
#include "highlighter.h"

editor_t e;

void AddCharacter(char ch) {
	if (e.__selection.GetSelectionString(e.code_str()).length() != 0) { Cut(false); }

	Focus();
	InsertCharacter(e.row, e.col, ch);
	e.col++;

	char val;
	bool found = MaybeAddPair(e.row, e.col, ch, &val);
	if (found) {
		InsertCharacter(e.row, e.col, val);
	}

	if (e.isContentChanged) {
		UpdateLsp(e.code_str(), false);
		FindTests();
	}
}

void InsertCharacter(int line, int pos, char ch) {
	int offset = LineOffset(e.code_str(), line) + pos;

	// adding one character offset after first line
	// this is required due to the newline characters in the string buffer
	if (line > 0) { offset += 1; }

	e.code.insert(offset, ch);
	InsertTextEdit(e.code_str(), offset, 1);

	// undo operation type is switched between insert and addchar types
	action_t op;

	// change op if this is a new line character
	if (ch == '\n') { op = INSERT; } else { op = ADDCHAR; }

	// record the operation on the undo stack. Note that we're creating a new EditOperation
	// and adding all the Operations to it
	e.undo.push_back({operation_t{op, std::string(1, ch), offset, cursormove_t{line, pos}}});
}

void InsertString(int line, int pos, std::string linestring) {
	// modify string to remove tabs and space from beginning
	std::string l = RemoveLeadingTabsSpaces(linestring);
	int offset = LineOffset(e.code_str(), line) + pos + 1;
	e.code.insert(offset, l.c_str());
	InsertTextEdit(e.code_str(), offset, l.length());

	// record the operation on the undo stack. Note that we're creating a new EditOperation
	// and adding all the Operations to it
	e.undo.push_back({operation_t{INSERT, l, offset, cursormove_t{line, pos}}});
}

void DeleteCharacter(int line, int pos) {
	int offset = LineOffset(e.code_str(), line) + pos;
	char ch = e.code.at(offset);
	e.code.erase(offset, 1);
	RemoveTextEdit(e.code_str(), offset, 1);

	// record the operation on the undo stack. Note that we're creating a new EditOperation
	// and adding all the Operations to it
	e.undo.push_back({operation_t{DELETE, std::string(1, ch), offset, cursormove_t{line, pos}}});
}

void ReplaceString(int line, int from, int end, std::string instext) {
	int offset = LineOffset(e.code_str(), line);
	int begin_idx = offset + from + 1;

	auto deltext = e.code.substr(begin_idx, end - from);
	auto deltext_str = std::string(deltext.begin(), deltext.end());
	auto instext_str = std::string(instext.begin(), instext.end());

	e.code.erase(begin_idx, deltext.size());
	RemoveTextEdit(e.code_str(), begin_idx, deltext.size());

	e.code.insert(begin_idx, instext.c_str());
	InsertTextEdit(e.code_str(), begin_idx, instext.size());

	// record the operation on the undo stack. Note that we're creating a new EditOperation
	// and adding all the Operations to it
	e.undo.push_back(editoperation_t{
		operation_t{DELETE, deltext_str, begin_idx, cursormove_t{line, from}},
		operation_t{INSERT, instext_str, begin_idx, cursormove_t{line, from}}
	});
}

void ShiftWithTabsToRight(int line, int pos, std::set<int> selectedLines) {
	e.__selection.ssx = 0;

	editoperation_t ops{};
	for (int linenumber : selectedLines) {
		int offset = LineOffset(e.code_str(), linenumber);
		e.code.insert(offset, "\t");
		ops.push_back(operation_t{INSERT, "\t", offset, cursormove_t{line, pos}});
	}

	e.__selection.sex = pos;
	e.undo.push_back(ops);
}

bool MaybeAddPair(int line, int pos, char ch, char *ret) {
	std::map<char, char> pairMap = {
		{'(', ')'}, {'{', '}'}, {'[', ']'}, {'"', '"'}, {'\\', '\\'}, {'`', '`'},
	};

	if (e.code.size() == 0) { *ret = '\0'; return false; }

	int offset_start = LineOffset(e.code_str(), line);
	int offset_end = LineOffset(e.code_str(), line + 1);
	auto line_str = e.code.substr(offset_start + 1, offset_end);
	auto current_char = e.code.at(offset_start + 1 + pos - 2);

	if (pairMap.count(ch)) {
		bool noMoreChars = pos >= (int)line_str.size() - 1;
		bool isSpaceNext = pos < (int)line_str.size() - 1 && current_char == ' ';
		bool isStringAndClosedBracketNext = pairMap[ch] == '"' && pos < (int)line_str.size() - 1 && current_char == ')';

		if (noMoreChars || isSpaceNext || isStringAndClosedBracketNext) {
			*ret = pairMap[ch]; // returning the character
			return true; // found character pair
		}
	}

	*ret = '\0'; // returning an empty character
	return false; // could not found
}
